lib/OpenQA/WebAPI/Plugin/Gru.pm

```diff
@@ -127,10 +127,12 @@ sub cmd_list { shift->job->run(@_) }
 
 sub execute_job {
     my ($self, $job) = @_;
-    unless (my $err = $job->_run) {
-        $job->finish;
-        $self->delete_gru($job->info->{notes}{gru_id}) if exists $job->info->{notes}{gru_id};
-    }
+
+    if   (my $err = $job->execute) { $job->fail($err) }
+    else                           { $job->finish }
+
+    # Delete always the corresponding gru task or we would get pending gru jobs that should be manually deleted.
+    $self->delete_gru($job->info->{notes}{gru_id}) if exists $job->info->{notes}{gru_id};
```

Comment:
Yeah, before the Minion switch failed GRU tasks were just kept in the table and always retried (succeded ones instead were deleted to unblock openQA jobs from being scheduled). 

When changing this line now i thought is more proper to delete gru tasks also if failed (so openQA jobs should end up as failed as well, instead of being kept in scheduled) - @coolo ?

But now I'm actually wondering of ```->fail(``` since we should have to check if retrials have been exhausted or not before deleting it (before _run was calling it without checking it's return). Please correct if i'm wrong, but afaics from Minion code ```fail()``` it is returning true if the job actually fails or undef if it is being re-attempted.

Reply:
The return value only tells you if the state transition was successful.

Reply:
Don't forget that any job can also be retied manually from the command line and admin ui. If that gru data is required for that to work, maybe you should keep it around longer.

Reply:
Maybe you can just get rid of the gru data and rely solely on Minion? Metadata can be stored in notes and to see if a job is pending you could just do something like:
```perl
my $job = $minion->job($job_id);
my $pending = !$job || $job->info->{state} eq 'finished';
```

Reply:
@kraih point is that openQA jobs that are waiting for a GRU task will be on hold from being assigned by the scheduler. That would mean OpenQA::Scheduler should be adapted to scrape from Minion results, and with it decouple a bit more data replicating it into Minion metadata.

Reply:
@coolo in such case a new Minion job would be created, and not the failed one re-attempted